barretenberg/cpp/src/barretenberg/commitment_schemes/shplonk/shplonk.hpp

```diff
@@ -206,67 +203,235 @@ template <typename Curve> class ShplonkVerifier_ {
                 commitments.emplace_back(commitment);
                 scalars.emplace_back(-scaling_factor);
             }
-
             commitments.emplace_back(g1_identity);
             scalars.emplace_back(G_commitment_constant);
-
             // [G] += G₀⋅[1] = [G] + (∑ⱼ ρʲ ⋅ vⱼ / ( r − xⱼ ))⋅[1]
             G_commitment = GroupElement::batch_mul(commitments, scalars);
-
         } else {
             // [G] = [Q] - ∑ⱼ ρʲ / ( r − xⱼ )⋅[fⱼ] + G₀⋅[1]
             //     = [Q] - [∑ⱼ ρʲ ⋅ ( fⱼ(X) − vⱼ) / ( r − xⱼ )]
             G_commitment = Q_commitment;
-
             // Compute {ẑⱼ(r)}ⱼ , where ẑⱼ(r) = 1/zⱼ(r) = 1/(r - xⱼ)
             std::vector<Fr> inverse_vanishing_evals;
             inverse_vanishing_evals.reserve(num_claims);
             for (const auto& claim : claims) {
                 inverse_vanishing_evals.emplace_back(z_challenge - claim.opening_pair.challenge);
             }
             Fr::batch_invert(inverse_vanishing_evals);
-
             auto current_nu = Fr(1);
             // Note: commitments and scalars vectors used only in recursion setting for batch mul
             for (size_t j = 0; j < num_claims; ++j) {
                 // (Cⱼ, xⱼ, vⱼ)
                 const auto& [opening_pair, commitment] = claims[j];
-
                 Fr scaling_factor = current_nu * inverse_vanishing_evals[j]; // = ρʲ / ( r − xⱼ )
-
                 // G₀ += ρʲ / ( r − xⱼ ) ⋅ vⱼ
                 G_commitment_constant += scaling_factor * opening_pair.evaluation;
-
                 // [G] -= ρʲ / ( r − xⱼ )⋅[fⱼ]
                 G_commitment -= commitment * scaling_factor;
-
                 current_nu *= nu;
             }
-
             // [G] += G₀⋅[1] = [G] + (∑ⱼ ρʲ ⋅ vⱼ / ( r − xⱼ ))⋅[1]
             G_commitment += g1_identity * G_commitment_constant;
         }
+        // Return opening pair (z, 0) and commitment [G]
+        return { { z_challenge, Fr(0) }, G_commitment };
+    };
+
+    /**
+    Shplonk verifier optimized to verify gemini opening claims.
+
+    This method receives commitments to all prover polynomials, their claimed evaluations, the sumcheck
+    challenge, a challenge \f$ \rho \f$ aimed to batch the commitments to prover polynomials, a challenge \f$ r \f$ for
+    the Gemini opening claims, and the Gemini claims. The latter is a tuple of a vector of powers of \f$ r \f$, a vector
+    of evaluations of Gemini fold polynomials at \f$ - r,
+    - r^2, \ldots, - r^{2^{d-1}} \f$ where \f$ d \f$ is the log_circuit_size, and a vector of commitments to the Gemini
+    fold polynomials.
+
+    The verifier receives the challenges required in Shplonk and gradually populates the vectors of commitments and
+    scalars that will be multiplied at the very end. In the recursive setting, a batch_mul of size (NUM_ALL_ENTITIES +
+    log_circuit_size + 2) is performed. In the native setting, these operations are performed sequentially.
+     *
+     */
+    static OpeningClaim<Curve> verify_gemini(
```

Comment:
Maybe for unifority we could rename this to something like `reduce_verification_for_gemini`?

Reply:
Also: it feels kind of wrong that so much has moved out of gemini and into here. It also makes it harder to understand the diff. If our implementation of Shplonk now depends on our implementation of Gemini, then perhaps it makes sense to leave more of this stuff in Gemini, keep around the original version for reference (also who knows if we'll want it...) and share code with Shplonk at the expense of including Gemini in Shplonk? If we really wanted, we could make a second moving-around PR where Shplemini is implemented in a third file and there's not Gemini <-> Shplonk dependency.